File: coding_skill_py/31_tracing_pattern.py
"""Python Tracing - 데코레이터"""
from functools import wraps
from typing import Callable

_trace_depth = 0

def trace(func: Callable) -> Callable:
    @wraps(func)
    def wrapper(*args, **kwargs):
        global _trace_depth
        print(" " * _trace_depth + f"→ {func.__name__}()")
        _trace_depth += 1
        try:
            result = func(*args, **kwargs)
            return result
        finally:
            _trace_depth -= 1
            print(" " * _trace_depth + f"← {func.__name__}()")
    return wrapper

@trace
def function_c():
    print("      FunctionC 실행")

@trace
def function_b():
    print("      FunctionB 실행")
    function_c()

@trace
def function_a():
    print("      FunctionA 실행")
    function_b()

def main():
    print("=== Python Tracing ===")
    function_a()

if __name__ == "__main__":
    main()

File: coding_skill_cpp/31_tracing_pattern.cpp
/* C++ Tracing - RAII 기반 */
#include <iostream>
#include <string>
using namespace std;

class FunctionTracer {
    string name;
    static int depth;
public:
    FunctionTracer(const string& fname) : name(fname) {
        cout << string(depth++, ' ') << "→ " << name << "()" << endl;
    }
    
    ~FunctionTracer() {
        cout << string(--depth, ' ') << "← " << name << "()" << endl;
    }
};

int FunctionTracer::depth = 0;

#define TRACE() FunctionTracer __tracer__(__func__)

void functionC() {
    TRACE();
    cout << "      FunctionC 실행" << endl;
}

void functionB() {
    TRACE();
    cout << "      FunctionB 실행" << endl;
    functionC();
}

void functionA() {
    TRACE();
    cout << "      FunctionA 실행" << endl;
    functionB();
}

int main() {
    cout << "=== C++ Tracing ===" << endl;
    functionA();
    return 0;
}
